configure.ac

```diff
@@ -499,16 +518,6 @@ AC_ARG_ENABLE([cuda-dlopen],
     ],
     [enable_cuda_dlopen=no])
 
-FI_CHECK_PACKAGE([cuda],
-		 [cuda_runtime.h],
-		 [cudart],
-		 [cudaMemcpy],
-		 [-lcuda],
-		 [$with_cuda],
-		 [],
-		 [AC_DEFINE([HAVE_LIBCUDA], [1],[CUDA support])],
-		 [], [])
```

Comment:
After the AC_DEFINE case (the next []), is there some way to see if with-cuda has been specified, and if so, call AC_MSG_ERROR here?  That would avoid duplicating the FI_CHECK_PACKAGE.  Maybe it requires checking AC_ARG_WITH again.  E.g.
```
[AC_DEFINE([HAVE_LIBCUDA], [1], [CUDA support])],
[AC_ARG_WITH([cuda],[AC_MSG_ERROR([CUDA support requested...])], [])], [])
```

Reply:
Two suggestions:

1. It is significantly better to not duplicate the whole call to `FI_CHECK_PACKAGE` twice.  
1. You should not `AC_DEFINE` inside a conditional execution to just 1 (and leave it undefined for the false case).  Instead, it is sigificantly safer to *always* `AC_DEFINE` the macro, and define it to 0 or 1.

Might I humbly suggest something like this (typed here in github -- not checked for errors/typos):

```
AC_ARG_WITH([cuda],
	    [AC_HELP_STRING([--with-cuda=DIR],
			    [Provide path to where the CUDA development
			    and runtime libraries are installed.])])

have_libcuda=0
AS_IF([test "$with_cuda" = "yes"],
     [FI_CHECK_PACKAGE([cuda],
		 [cuda_runtime.h],
		 [cudart],
		 [cudaMemcpy],
		 [-lcuda],
		 [$with_cuda],
		 [],
		 [have_libcuda=1])
AS_IF([test "$with_cuda" = "yes" && $have_libcuda -eq 0],
    [AC_MSG_ERROR([CUDA support requested but CUDA runtime not available.])])
AC_DEFINE_UNQUOTED([HAVE_LIBCUDA], [$have_libcuda], [Whether we have libcuda or not])
```

Because this:

```c
#if HAVE_LIBCUDA
   // ...
#endif
```

is safer / will cause compiler errors if you typo the macro name anywhere, vs.:

```c
#if defined(HAVE_LIBCUDA)
   // ...
#endif
```

which won't.

Reply:
✅